File: java/1905-count-sub-islands.java
class Solution {
    class RecursiveBiFunction<A, B, C> {
        BiFunction<A, B, C> func;
    }
    
    public int countSubIslands(final int[][] grid1, final int[][] grid2) {
        final int ROWS = grid1.length, COLS = grid1[0].length;
        final Set<Integer> visit = new HashSet<>();
        
        final RecursiveBiFunction<Integer, Integer, Boolean> dfs = new RecursiveBiFunction<>();
        dfs.func = (r, c) -> {
            int flatCoord = r*COLS + c;
            if(
                r < 0 
                || c < 0
                || r == ROWS
                || c == COLS
                || grid2[r][c] == 0
                || visit.contains(flatCoord)
            )
                return true;
            
            visit.add(flatCoord);
            boolean res = true;
            if(grid1[r][c] == 0)
                res = false;
            
            res = dfs.func.apply(r - 1, c) && res;
            res = dfs.func.apply(r + 1, c) && res;
            res = dfs.func.apply(r, c - 1) && res;
            res = dfs.func.apply(r, c + 1) && res;
            return res;
        };
        
        int count = 0;
        for(int r = 0; r < ROWS; r++)
            for(int c = 0; c < COLS; c++)
                if(grid2[r][c] != 0 && !visit.contains(r*COLS + c) && dfs.func.apply(r, c))
                    count += 1;
        return count;
    }
}

File: javascript/1905-count-sub-islands.js
var countSubIslands = function(grid1, grid2) {
    let ROWS = grid1.length, COLS = grid1[0].length;
    let visit = new Set();
    
    const dfs = function(r, c) {
        let flatCoord = r*COLS + c;
        if (
            r < 0
            || c < 0
            || r == ROWS
            || c == COLS
            || grid2[r][c] == 0
            || visit.has(flatCoord)
        )
            return true;
        
        visit.add(flatCoord);
        let res = true;
        if(grid1[r][c] == 0)
            res = false;
        
        res = dfs(r - 1, c) && res;
        res = dfs(r + 1, c) && res;
        res = dfs(r, c - 1) && res;
        res = dfs(r, c + 1) && res;
        return res;
    };
    
    let count = 0;
    for(let r = 0; r < ROWS; r++)
        for(let c = 0; c < COLS; c++)
            if(grid2[r][c] && !visit.has(r*COLS + c) && dfs(r, c))
                count += 1;
    return count;
};
